src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,378 @@
+            s_optionsLength[(int)PreLoginOptions.MARS] = 1; // MARS
+            s_optionsLength[(int)PreLoginOptions.TRACEID] = GUID_SIZE + GUID_SIZE + sizeof(uint); // TRACEID
+            s_optionsLength[(int)PreLoginOptions.FEDAUTHREQUIRED] = 1; // FEDAUTHREQUIRED
+            
+            // Calculate the payload length.
+            for (int i = 0; i < (int)PreLoginOptions.NUMOPT; i++)
+            {
+                s_optionsPayloadLength += s_optionsLength[i];
+            }
+        }
+
+        // EventSource counter
+        private static int s_objectTypeCount;
+
+        internal readonly int _objectID = Interlocked.Increment(ref s_objectTypeCount);
+
+        internal int ObjectID => _objectID;
+
+        public IHandler<PreloginHandlerContext> NextHandler { get; set; }
+
+        public async ValueTask Handle(PreloginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            await PreloginPacketHandler.CreatePreloginAndSend(context, isAsync, ct).ConfigureAwait(false);
+            
+            await ReadPreloginResponse(context, isAsync, ct).ConfigureAwait(false);
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Consume the Prelogin response from the server.
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async Task ReadPreloginResponse(PreloginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            context.ConnectionContext.MarsCapable = context.ConnectionContext.ConnectionString.MARS; // Assign default value
+            context.ConnectionContext.FedAuthRequired = false;
+
+            TdsStream tdsStream = context.ConnectionContext.TdsStream;
+
+            // Force a read on the TDS stream;
+            _ = await tdsStream.PeekByteAsync(isAsync, ct).ConfigureAwait(false);
+            
+            byte[] preloginPayload = new byte[tdsStream.PacketDataLeft];
+            if (isAsync)
+            {
+                _ = await tdsStream.ReadAsync(preloginPayload.AsMemory(), ct).ConfigureAwait(false);
+            }
+            else
+            {
+                tdsStream.Read(preloginPayload);
+            }
+
+            int payloadOffset = 0;
+            int offset = 0;
+            PreLoginOptions option = (PreLoginOptions)preloginPayload[offset++];
+            while (option != PreLoginOptions.LASTOPT)
+            {
+                switch (option)
+                {
+                    case PreLoginOptions.VERSION:
+                        // Nothing to do with the version.
+                        offset += PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES; // Skip the payload offset and length
```

Comment:
If this is where we're going to verify/block SQL Server versions, it's probably worth adding a TODO here.

Reply:
No we wont be using this anymore. We use the login ack to get the server version now, and any feature negotiations between client and server.

Reply:
This would work, but seems a little late - is there an observed difference between the version presented in the pre-login and login ack messages for some products? By the time we receive a login ack message, I'd expect the client to have authenticated to the server, having incurred the overhead (and possible exceptions) of a Kerberos token retrieval, AAD access token generation, etc. I could definitely understand it if some editions of SQL Server didn't want to publish their version number to an unauthenticated client though...

A feature negotiation dynamically permits the client to use certain SqlClient features, but the blocks on the server version will be constant for a given version of SqlClient. If the pre-login message's version field is trustworthy then I think there's a case for treating these slightly differently and blocking before a login takes place.

Reply:
>  is there an observed difference between the version presented in the pre-login and login ack messages for some products? 

No idea. 

In the initial versions of the protocol pre-login used to exchange the server version, which seemed to have been limited to saying whether the server supports MARS or not.
Then came login ack for version exchange so that the server version can be used to do feature switches, along with protocol versions. 
Then came Azure, where the version doesn't matter at all. And hence all negotiations happen after login, to decouple features from being tied to server. Azure SQL DB and Azure SQL MI are version less. 

Hence a client can support a feature on Azure if it's available (usually Azure gets most of the features before they are shipped on-prem). And when the feature comes to the On-prem server, then we know that the client's feature negotiation capability can kick in. 

Then there are some features which are definitely not protocol negotiable, like TDS8.0 vs 7.4 where TLS handshake is impacted, which needs to be done without connecting to the server, via Encrypt parameter in the connection string. 

> By the time we receive a login ack message, I'd expect the client to have authenticated to the server, having incurred the overhead (and possible exceptions) of a Kerberos token retrieval, AAD access token generation, etc. I could definitely understand it if some editions of SQL Server didn't want to publish their version number to an unauthenticated client though...

The overhead of negotiating Kerberos / AAD token etc, is orthogonal to feature negotiation. Whether is feature is negotiated before successful login or after, they are pointless if the login doesn't work, because there wont be a connection in case of a failed login to exercise a feature. 

Where in the connectivity flow, we identify the logical feature, really doesn't matter in the bigger scheme of things, as long as it is somewhere in the connectivity flow. 

I can see that the driver stores the fact that it is connected to SQL 2022, but that really is not being used for any decisions anymore, it's the feature extensions ack that matters for versions of SQL on-prem released since 2016. There i no sql2016/2019 in the list either.

Reply:
I completely agree about feature negotiation: it should happen in the login ack, we should be using feature negotiation rather than assuming a specific version supports a feature, and the authentication overhead's definitely a separate concern.

My question is separate to feature negotiation though. SqlClient and SqlClientX will explicitly block unsupported versions of SQL Server (SQL 2014 and below, by the time SqlClientX is released.) I expected that we would look at the pre-login message to perform this block in SqlClientX, so the library could throw an exception for the developer. I'm asking whether there's an MS-internal reason why we can't perform this block in the pre-login message; I'd prefer to throw the exception early and avoid the Kerberos / AAD authentication overhead which would be incurred if we waited until the login-ack message.

Reply:
I get your point now. The intention is to block connectivity to unsupported SQL Server versions. 

@edwardneal traditionally we have never explicitly blocked on unsupported versions of SQL server. Doing this creates an expectation that a new client will always be updated when a server version is deprecated else we run into issues like, "the client explicitly says that it doesn't work for SQL Server 2005, but is not blocking for SQL 2xxx" does this mean that the client support 2xxx, which the docs say is not - supported? 
The way we have always approached not supporting SQL server versions is to not address any issues which come up due to connectivity / query executions to those versions. The statement is `if client/server works for an unsupported client/server version, and you are using the client/server, then do it at your own risk`. No servicing issues will be fixed in older drivers, and certainly not addressed in the latest code of the driver.

Removing the checks in SqlClient, for unsupported version is good enough reason for a user to not use these server versions.

Reply:
That's fair enough. I was working from [this comment](https://github.com/dotnet/SqlClient/pull/2601#discussion_r1655478850):
> There's no reason that main needs to work on anything older than 2014. Scratch that. 2016, since 2014 EOLs July 9. Users who are running EOL servers can use old clients.

which was in response to an earlier question about whether the support policy should be enforced in the main branch.

I can see that later in the thread you talk about opening a separate issue to talk about the compatibility for the rest of the driver. That's probably a better place to talk about this than this PR - it's definitely not holding its merge back!

Reply:
Thanks @edwardneal